pandas/io/stata.py

```diff
@@ -1521,42 +1516,64 @@ def _read_value_labels(self) -> None:
             self._path_or_buf.seek(self._data_location + offset)
 
         self._value_labels_read = True
-        self._value_label_dict = {}
+        self._value_label_dict: dict[str, dict[int, str]] = {}
 
         while True:
             if self._format_version >= 117:
                 if self._path_or_buf.read(5) == b"</val":  # <lbl>
                     break  # end of value label table
 
-            slength = self._path_or_buf.read(4)
-            if not slength:
-                break  # end of value label table (format < 117)
-            if self._format_version <= 117:
-                labname = self._decode(self._path_or_buf.read(33))
+            if self._format_version >= 108:
```

Comment:
Something is off with the logic here and a few lines above with the other check. This should only be read if < 117 it looks like.

Reply:
The content of `if self._format_version >= 108:` are almost identical to the original code (though the change in indentation makes this less obvious in the diff). All that I changed within the if section is to add the 9 character read when version is 108. The new code is mostly in the else statement.

I hadn't looked at the slength bit previously, but it could probably be a bit clearer. The `if self._format_version >= 117:` lines should mean that for a well-formed file the loop breaks before `if not slength:` becomes true, however for a case where `</val` is missing or in the wrong place the while loop would not exit if I added `< 117` to the `is not slength` condition. Basically, as I understand it, this is allowing the loop to end if an end-of-file is reached. In versions < 117 this is how the end of the value label section is indicated, but in 117 and greater this should be indicted with the string </value_labels>.